pkg/apis/pipeline/v1alpha1/volume_resource_test.go

```diff
@@ -0,0 +1,158 @@
+	}{{
+		desc: "basic volume resource",
+		resource: tb.PipelineResource("test-volume-resource", "default", tb.PipelineResourceSpec(
+			v1alpha1.PipelineResourceTypeVolume,
+			tb.PipelineResourceSpecParam("name", "test-volume-resource"),
+		)),
+		want: &v1alpha1.VolumeResource{
+			Name: "test-volume-resource",
+			Type: v1alpha1.PipelineResourceTypeVolume,
+			Size: "5Gi",
+		},
+	}, {
+		desc: "volume resource with size",
+		resource: tb.PipelineResource("test-volume-resource", "default", tb.PipelineResourceSpec(
+			v1alpha1.PipelineResourceTypeVolume,
+			tb.PipelineResourceSpecParam("name", "test-volume-resource"),
+			tb.PipelineResourceSpecParam("size", "10Gi"),
+		)),
+		want: &v1alpha1.VolumeResource{
+			Name: "test-volume-resource",
+			Type: v1alpha1.PipelineResourceTypeVolume,
+			Size: "10Gi",
+		},
+	}} {
+		t.Run(c.desc, func(t *testing.T) {
+			got, err := v1alpha1.NewVolumeResource(c.resource)
+			if err != nil {
+				t.Errorf("Test: %q; TestNewVolumeResource() error = %v", c.desc, err)
+			}
+			if d := cmp.Diff(got, c.want); d != "" {
+				t.Errorf("Diff:\n%s", d)
+			}
+		})
+	}
+}
+
+func Test_VolumeResource_GetDownloadContainerSpec(t *testing.T) {
+	names.TestingSeed()
+	testcases := []struct {
+		name           string
+		volumeResource *v1alpha1.VolumeResource
+		wantContainers []corev1.Container
+		wantErr        bool
+	}{{
+		name: "valid volume resource config",
+		volumeResource: &v1alpha1.VolumeResource{
+			Name:           "test-volume-resource",
+			Type:           v1alpha1.PipelineResourceTypeVolume,
+			Size:           v1alpha1.DefaultPvcSize,
+			DestinationDir: "/workspace",
+			SourceDir:      "/src-dir",
+		},
+		wantContainers: []corev1.Container{{
+			Name:    "create-dir-test-volume-resource-9l9zj",
+			Image:   "override-with-bash-noop:latest",
+			Command: []string{"/ko-app/bash"},
+			Args:    []string{"-args", "mkdir -p /workspace"},
+		}, {
+			Name:    "download-copy-test-volume-resource-mz4c7",
+			Image:   "override-with-bash-noop:latest",
+			Command: []string{"/ko-app/bash"},
+			Args:    []string{"-args", "cp -r /volumeresource/src-dir/. /workspace"},
+			VolumeMounts: []corev1.VolumeMount{{
+				Name:      "test-volume-resource",
+				MountPath: "/volumeresource",
+			}},
+		}},
+	}}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			gotContainers, err := tc.volumeResource.GetDownloadContainerSpec()
+			if tc.wantErr && err == nil {
+				t.Fatalf("Expected error to be %t but got %v:", tc.wantErr, err)
+			}
+			if d := cmp.Diff(gotContainers, tc.wantContainers); d != "" {
+				t.Errorf("Error mismatch between download containers spec: %s", d)
+			}
+		})
+	}
+}
+
+func Test_VolumeResource_GetUploadContainerSpec(t *testing.T) {
+	names.TestingSeed()
+	testcases := []struct {
+		name           string
+		volumeResource *v1alpha1.VolumeResource
+		wantContainers []corev1.Container
+		wantErr        bool
+	}{{
+		name: "valid volume resource config",
```

Comment:
in cases where we only have one test case, the test might be a bit easier to read without the table driven logic